File: Sources/FWObjC/Service/Request/ChainRequest.m
//
//  ChainRequest.m
//
//  Copyright (c) 2012-2016 FWNetwork https://github.com/yuantiku
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.

#import "ChainRequest.h"
#import "RequestConfig.h"
#import "BaseRequest.h"
#import "RequestManager.h"
#import "ObjC.h"

@interface __FWChainRequest()<__FWRequestDelegate>

@property (strong, nonatomic) NSMutableArray<__FWBaseRequest *> *requestArray;
@property (strong, nonatomic) NSMutableArray<__FWChainCallback> *requestCallbackArray;
@property (assign, nonatomic) NSUInteger nextRequestIndex;
@property (weak, nonatomic) __FWBaseRequest *nextRequest;
@property (copy, nonatomic) __FWChainCallback emptyCallback;

@end

@implementation __FWChainRequest

- (instancetype)init {
    self = [super init];
    if (self) {
        _nextRequestIndex = 0;
        _requestArray = [NSMutableArray array];
        _requestCallbackArray = [NSMutableArray array];
        _stoppedOnFailure = YES;
        _emptyCallback = ^(__FWChainRequest *chainRequest, __FWBaseRequest *baseRequest) { };
    }
    return self;
}

- (void)start {
    if (_nextRequestIndex > 0) {
        if ([__FWRequestConfig sharedConfig].debugLogEnabled) {
            FWLogDebug(@"Error! Chain request has already started.");
        }
        return;
    }

    _succeedRequest = nil;
    _failedRequest = nil;
    [[__FWRequestManager sharedManager] addChainRequest:self];
    [self toggleAccessoriesWillStartCallBack];
    if (![self startNextRequest:nil]) {
        if ([__FWRequestConfig sharedConfig].debugLogEnabled) {
            FWLogDebug(@"Error! Chain request array is empty.");
        }
    }
}

- (void)stop {
    [self toggleAccessoriesWillStopCallBack];
    _delegate = nil;
    [self clearRequest];
    [self toggleAccessoriesDidStopCallBack];
    [[__FWRequestManager sharedManager] removeChainRequest:self];
}

- (void)startWithSuccess:(void (^)(__FWChainRequest *chainRequest))success
                 failure:(void (^)(__FWChainRequest *chainRequest))failure {
    [self setCompletionBlockWithSuccess:success failure:failure];
    [self start];
}

- (void)startWithCompletion:(void (^)(__FWChainRequest *))completion {
    [self startWithSuccess:completion failure:completion];
}

- (void)startWithWillStart:(nullable void (^)(__FWChainRequest *chainRequest))willStart
                  willStop:(nullable void (^)(__FWChainRequest *chainRequest))willStop
                   success:(nullable void (^)(__FWChainRequest *chainRequest))success
                   failure:(nullable void (^)(__FWChainRequest *chainRequest))failure
                   didStop:(nullable void (^)(__FWChainRequest *chainRequest))didStop {
    __FWRequestAccessory *accessory = [__FWRequestAccessory new];
    accessory.willStartBlock = willStart;
    accessory.willStopBlock = willStop;
    accessory.didStopBlock = didStop;
    [self addAccessory:accessory];
    [self startWithSuccess:success failure:failure];
}

- (void)startSynchronouslyWithSuccess:(void (^)(__FWChainRequest * _Nonnull))success failure:(void (^)(__FWChainRequest * _Nonnull))failure {
    [self startSynchronouslyWithFilter:nil completion:^(__FWChainRequest *chainRequest) {
        if (chainRequest.failedRequest == nil) {
            if (success) success(chainRequest);
        } else {
            if (failure) failure(chainRequest);
        }
    }];
}

- (void)startSynchronouslyWithFilter:(BOOL (^)(void))filter completion:(void (^)(__FWChainRequest * _Nonnull))completion {
    [[__FWRequestManager sharedManager] synchronousChainRequest:self filter:filter completion:completion];
}

- (void)toggleAccessoriesWillStartCallBack {
    for (id<__FWRequestAccessory> accessory in self.requestAccessories) {
        if ([accessory respondsToSelector:@selector(requestWillStart:)]) {
            [accessory requestWillStart:self];
        }
    }
}

- (void)toggleAccessoriesWillStopCallBack {
    for (id<__FWRequestAccessory> accessory in self.requestAccessories) {
        if ([accessory respondsToSelector:@selector(requestWillStop:)]) {
            [accessory requestWillStop:self];
        }
    }
}

- (void)toggleAccessoriesDidStopCallBack {
    for (id<__FWRequestAccessory> accessory in self.requestAccessories) {
        if ([accessory respondsToSelector:@selector(requestDidStop:)]) {
            [accessory requestDidStop:self];
        }
    }
}

- (void)setCompletionBlockWithSuccess:(void (^)(__FWChainRequest *chainRequest))success
                              failure:(void (^)(__FWChainRequest *chainRequest))failure {
    self.successCompletionBlock = success;
    self.failureCompletionBlock = failure;
}

- (void)clearCompletionBlock {
    // nil out to break the retain cycle.
    self.successCompletionBlock = nil;
    self.failureCompletionBlock = nil;
}

- (void)dealloc {
    [self clearRequest];
}

- (void)addRequest:(__FWBaseRequest *)request callback:(__FWChainCallback)callback {
    [_requestArray addObject:request];
    if (callback != nil) {
        [_requestCallbackArray addObject:callback];
    } else {
        [_requestCallbackArray addObject:_emptyCallback];
    }
}

- (NSArray<__FWBaseRequest *> *)requestArray {
    return _requestArray;
}

- (BOOL)startNextRequest:(__FWBaseRequest *)previousRequest {
    if (_nextRequestIndex >= [_requestArray count] && self.requestBuilder) {
        __FWBaseRequest *baseRequest = self.requestBuilder(self, previousRequest);
        if (baseRequest) {
            [self addRequest:baseRequest callback:nil];
        }
    }
    
    if (_nextRequestIndex < [_requestArray count]) {
        __FWBaseRequest *request = _requestArray[_nextRequestIndex];
        _nextRequestIndex++;
        request.delegate = self;
        [request clearCompletionBlock];
        if (_nextRequestIndex > 1 && self.requestInterval > 0) {
            _nextRequest = request;
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(self.requestInterval * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [self.nextRequest start];
            });
        } else {
            _nextRequest = nil;
            [request start];
        }
        return YES;
    } else {
        return NO;
    }
}

#pragma mark - Network Request Delegate

- (void)requestFinished:(__FWBaseRequest *)request {
    _succeedRequest = request;
    _failedRequest = nil;
    
    NSUInteger currentRequestIndex = _nextRequestIndex - 1;
    __FWChainCallback chainCallback = _requestCallbackArray[currentRequestIndex];
    chainCallback(self, request);
    
    if (self.stoppedOnSuccess || ![self startNextRequest:request]) {
        [self requestCompleted];
    }
}

- (void)requestFailed:(__FWBaseRequest *)request {
    _succeedRequest = nil;
    _failedRequest = request;
    
    if (self.stoppedOnFailure || ![self startNextRequest:request]) {
        [self requestCompleted];
    }
}

- (void)requestCompleted {
    [self toggleAccessoriesWillStopCallBack];
    
    if (!_failedRequest) {
        if ([_delegate respondsToSelector:@selector(chainRequestFinished:)]) {
            [_delegate chainRequestFinished:self];
        }
        if (_successCompletionBlock) {
            _successCompletionBlock(self);
        }
    } else {
        if ([_delegate respondsToSelector:@selector(chainRequestFailed:)]) {
            [_delegate chainRequestFailed:self];
        }
        if (_failureCompletionBlock) {
            _failureCompletionBlock(self);
        }
    }
    
    [self clearCompletionBlock];
    [self toggleAccessoriesDidStopCallBack];
    [[__FWRequestManager sharedManager] removeChainRequest:self];
}

- (void)clearRequest {
    if (_nextRequestIndex > 0) {
        NSUInteger currentRequestIndex = _nextRequestIndex - 1;
        if (currentRequestIndex < [_requestArray count]) {
            __FWBaseRequest *request = _requestArray[currentRequestIndex];
            [request stop];
        }
    }
    _nextRequest = nil;
    [_requestArray removeAllObjects];
    [_requestCallbackArray removeAllObjects];
    _requestBuilder = nil;
    [self clearCompletionBlock];
}

#pragma mark - Request Accessoies

- (void)addAccessory:(id<__FWRequestAccessory>)accessory {
    if (!self.requestAccessories) {
        self.requestAccessories = [NSMutableArray array];
    }
    [self.requestAccessories addObject:accessory];
}

@end

File: Sources/FWFramework/Service/Request/ChainRequest.swift
//
//  ChainRequest.swift
//  FWFramework
//
//  Created by wuyong on 2023/11/17.
//

import Foundation

// MARK: - ChainRequest
/// 队列请求代理
public protocol ChainRequestDelegate: AnyObject {
    /// 队列请求完成
    @MainActor func chainRequestFinished(_ chainRequest: ChainRequest)
    /// 队列请求失败
    @MainActor func chainRequestFailed(_ chainRequest: ChainRequest)
}

extension ChainRequestDelegate {
    /// 默认实现队列请求完成
    public func chainRequestFinished(_ chainRequest: ChainRequest) {}
    /// 默认实现队列请求失败
    public func chainRequestFailed(_ chainRequest: ChainRequest) {}
}

/// 队列请求类
open class ChainRequest: HTTPRequestProtocol, RequestDelegate, @unchecked Sendable {
    /// 队列请求完成句柄
    public typealias Completion = @MainActor @Sendable (ChainRequest) -> Void
    /// 回调处理句柄声明
    public typealias CallbackHandler = @Sendable (ChainRequest, HTTPRequest) -> Void

    // MARK: - Accessor
    /// 当前请求数组
    open private(set) var requestArray: [HTTPRequest] = []
    /// 事件代理
    open weak var delegate: ChainRequestDelegate?
    /// 成功完成回调
    open var successCompletionBlock: Completion?
    /// 失败完成回调
    open var failureCompletionBlock: Completion?
    /// 请求结束回调，成功失败都会触发
    open var requestCompletedBlock: Completion?
    /// 请求取消回调，不一定主线程调用
    open var requestCancelledBlock: (@Sendable (ChainRequest) -> Void)?
    /// 请求标签，默认0
    open var tag: Int = 0
    /// 自定义请求配件数组
    open var requestAccessories: [RequestAccessoryProtocol]?
    /// 最后一个导致队列请求成功的请求
    open private(set) var succeedRequest: HTTPRequest?
    /// 最后一个导致队列请求失败的请求
    open private(set) var failedRequest: HTTPRequest?
    /// 是否自动显示错误信息
    open var autoShowError: Bool {
        get { failedRequest?.autoShowError ?? false }
        set { failedRequest?.autoShowError = newValue }
    }

    /// 当前网络错误
    open var error: Error? {
        failedRequest?.error
    }

    /// 请求是否已取消
    open private(set) var isCancelled = false
    /// 请求间的时间间隔
    open var requestInterval: TimeInterval = 0
    /// 某个请求失败时，是否立即停止队列请求，默认true
    open var stoppedOnFailure = true
    /// 某个请求成功时，是否立即停止队列请求，默认false
    open var stoppedOnSuccess = false
    /// 请求构建句柄，所有请求完成后才会主线程调用
    open var requestBuilder: (@Sendable (_ chainRequest: ChainRequest, _ previousRequest: HTTPRequest?) -> HTTPRequest?)?

    private var requestCallbackArray: [CallbackHandler] = []
    private var nextRequestIndex: Int = 0
    private weak var nextRequest: HTTPRequest?
    private let emptyCallback: CallbackHandler = { _, _ in }

    // MARK: - Lifecycle
    public init() {}

    deinit {
        clearRequest()
    }

    // MARK: - Public
    /// 添加请求，可设置请求完成回调
    @discardableResult
    open func addRequest(_ request: HTTPRequest, callback: CallbackHandler? = nil) -> Self {
        requestArray.append(request)
        requestCallbackArray.append(callback ?? emptyCallback)
        return self
    }

    /// 开始请求，仅能调用一次
    @discardableResult
    open func start() -> Self {
        guard nextRequestIndex <= 0 else { return self }

        succeedRequest = nil
        failedRequest = nil
        ChainRequestManager.shared.addChainRequest(self)
        startNextRequest(nil)
        return self
    }

    /// 取消请求
    open func cancel() {
        toggleAccessoriesWillStopCallBack()
        delegate = nil
        clearRequest()
        isCancelled = true
        requestCancelledBlock?(self)
        requestCancelledBlock = nil
        toggleAccessoriesDidStopCallBack()
        ChainRequestManager.shared.removeChainRequest(self)
    }

    /// 开始请求并指定成功、失败、结束句柄
    @discardableResult
    open func start(
        success: Completion?,
        failure: Completion?,
        complete: Completion? = nil
    ) -> Self {
        successCompletionBlock = success
        failureCompletionBlock = failure
        if complete != nil { requestCompletedBlock = complete }
        return start()
    }

    /// 开始请求并指定完成句柄
    @discardableResult
    open func start(completion: Completion?) -> Self {
        start(success: completion, failure: completion)
    }

    /// 自定义请求结束句柄，成功失败都会触发
    @discardableResult
    open func requestCompleted(_ block: Completion?) -> Self {
        requestCompletedBlock = block
        return self
    }

    /// 请求取消句柄，不一定主线程调用
    @discardableResult
    open func requestCancelled(_ block: (@Sendable (ChainRequest) -> Void)?) -> Self {
        requestCancelledBlock = block
        return self
    }

    /// 自定义响应完成句柄
    @discardableResult
    open func response(_ completion: Completion?) -> Self {
        responseSuccess(completion).responseFailure(completion)
    }

    /// 自定义响应成功句柄
    @discardableResult
    open func responseSuccess(_ block: Completion?) -> Self {
        successCompletionBlock = block
        return self
    }

    /// 自定义响应失败句柄
    @discardableResult
    open func responseFailure(_ block: Completion?) -> Self {
        failureCompletionBlock = block
        return self
    }

    /// 快捷设置响应失败句柄
    @discardableResult
    open func responseError(_ block: (@MainActor @Sendable (Error) -> Void)?) -> Self {
        failureCompletionBlock = { request in
            block?(request.error ?? RequestError.unknown)
        }
        return self
    }

    /// 清理完成句柄
    open func clearCompletionBlock() {
        successCompletionBlock = nil
        failureCompletionBlock = nil
        requestCompletedBlock = nil
    }

    /// 显示网络错误，默认显示Toast提示
    open func showError() {
        failedRequest?.showError()
    }

    /// 添加请求配件
    @discardableResult
    open func addAccessory(_ accessory: RequestAccessoryProtocol) -> Self {
        if requestAccessories == nil {
            requestAccessories = []
        }
        requestAccessories?.append(accessory)
        return self
    }

    /// 请求完成回调
    open func requestFinished(_ request: HTTPRequest) {
        succeedRequest = request
        failedRequest = nil

        let currentRequestIndex = nextRequestIndex - 1
        let chainCallback = requestCallbackArray[currentRequestIndex]
        chainCallback(self, request)

        if stoppedOnSuccess || !startNextRequest(request) {
            requestCompleted()
        }
    }

    /// 请求失败回调
    open func requestFailed(_ request: HTTPRequest) {
        succeedRequest = nil
        failedRequest = request

        if stoppedOnFailure || !startNextRequest(request) {
            requestCompleted()
        }
    }

    // MARK: - Private
    @discardableResult
    private func startNextRequest(_ previousRequest: HTTPRequest?) -> Bool {
        if nextRequestIndex >= requestArray.count, requestBuilder != nil {
            if let request = requestBuilder?(self, previousRequest) {
                addRequest(request, callback: nil)
            }
        }

        if previousRequest == nil {
            toggleAccessoriesWillStartCallBack()
        }

        if nextRequestIndex < requestArray.count {
            let request = requestArray[nextRequestIndex]
            nextRequestIndex += 1
            request.autoShowLoading = false
            request.autoShowError = false
            request.preloadCacheModel = false
            request.delegate = self
            request.clearCompletionBlock()
            if nextRequestIndex > 1 && requestInterval > 0 {
                nextRequest = request
                DispatchQueue.main.asyncAfter(deadline: .now() + requestInterval) {
                    self.nextRequest?.start()
                }
            } else {
                nextRequest = nil
                request.start()
            }
            return true
        }
        return false
    }

    private func toggleAccessoriesWillStartCallBack() {
        requestAccessories?.forEach { accessory in
            accessory.requestWillStart(self)
        }
    }

    private func toggleAccessoriesWillStopCallBack() {
        requestAccessories?.forEach { accessory in
            accessory.requestWillStop(self)
        }
    }

    private func toggleAccessoriesDidStopCallBack() {
        requestAccessories?.forEach { accessory in
            accessory.requestDidStop(self)
        }
    }

    @MainActor private func requestCompleted() {
        toggleAccessoriesWillStopCallBack()

        if failedRequest == nil {
            delegate?.chainRequestFinished(self)
            successCompletionBlock?(self)
            requestCompletedBlock?(self)
        } else {
            delegate?.chainRequestFailed(self)
            failureCompletionBlock?(self)
            requestCompletedBlock?(self)
        }

        clearCompletionBlock()
        toggleAccessoriesDidStopCallBack()
        ChainRequestManager.shared.removeChainRequest(self)
    }

    private func clearRequest() {
        if nextRequestIndex > 0 {
            let currentRequestIndex = nextRequestIndex - 1
            if currentRequestIndex < requestArray.count {
                let request = requestArray[currentRequestIndex]
                request.cancel()
            }
        }

        nextRequest = nil
        requestArray.removeAll()
        requestCallbackArray.removeAll()
        requestBuilder = nil
        clearCompletionBlock()
    }
}

// MARK: - ChainRequestManager
/// 队列请求管理器
open class ChainRequestManager: @unchecked Sendable {
    public static let shared = ChainRequestManager()

    private var chainRequestArray: [ChainRequest] = []
    private var lock = NSLock()

    public init() {}

    /// 添加队列请求
    open func addChainRequest(_ chainRequest: ChainRequest) {
        lock.lock()
        defer { lock.unlock() }
        chainRequestArray.append(chainRequest)
    }

    /// 移除队列请求
    open func removeChainRequest(_ chainRequest: ChainRequest) {
        lock.lock()
        defer { lock.unlock() }
        chainRequestArray.removeAll(where: { $0 === chainRequest })
    }
}

// MARK: - Concurrency+ChainRequest
extension ChainRequest {
    /// 异步获取完成响应，注意非Task取消也会触发(Continuation流程)
    public func response() async -> ChainRequest {
        await withTaskCancellationHandler {
            await withCheckedContinuation { continuation in
                requestCancelled { request in
                    if !Task.isCancelled {
                        continuation.resume(returning: request)
                    }
                }
                .response { request in
                    continuation.resume(returning: request)
                }
                .start()
            }
        } onCancel: {
            self.cancel()
        }
    }

    /// 异步获取成功响应，注意非Task取消也会触发(Continuation流程)
    public func responseSuccess() async throws -> ChainRequest {
        try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                requestCancelled { _ in
                    if !Task.isCancelled {
                        continuation.resume(throwing: CancellationError())
                    }
                }
                .responseSuccess { request in
                    continuation.resume(returning: request)
                }
                .responseError { error in
                    continuation.resume(throwing: error)
                }
                .start()
            }
        } onCancel: {
            self.cancel()
        }
    }
}
